File: php/src/Persistence/SeedLoader.php
<?php

declare(strict_types=1);

namespace MintWiki\Persistence;

use PDO;
use RuntimeException;

/**
 * Portable seed fixture 로더.
 *
 * ANSI SQL 기반의 INSERT 문을 파싱하여 데이터베이스에 로드한다.
 * PostgreSQL과 MariaDB 양쪽에서 동일하게 동작한다.
 */
class SeedLoader
{
    private string $fixturesDir;

    /**
     * 초기화.
     *
     * @param string|null $fixturesDir seed 파일들이 있는 디렉토리.
     *                                   지정하지 않으면 php/tests/fixtures/seed를 사용.
     */
    public function __construct(?string $fixturesDir = null)
    {
        if ($fixturesDir === null) {
            // 기본값: php/tests/fixtures/seed
            // 이 클래스가 php/src/Persistence에 있으므로 상대 경로로 접근
            $projectRoot = dirname(__DIR__, 2);
            $fixturesDir = $projectRoot . '/tests/fixtures/seed';
        }

        if (!is_dir($fixturesDir)) {
            throw new RuntimeException(
                "Fixtures 디렉토리를 찾을 수 없습니다: {$fixturesDir}"
            );
        }

        $this->fixturesDir = $fixturesDir;
    }

    /**
     * 특정 테이블의 seed 데이터를 로드한다.
     *
     * @param PDO $connection PDO 연결
     * @param string $tableName 로드할 테이블명 (파일명으로도 사용: {table_name}.sql)
     */
    public function loadSeed(PDO $connection, string $tableName): void
    {
        $sqlFile = $this->fixturesDir . '/' . $tableName . '.sql';

        if (!file_exists($sqlFile)) {
            throw new RuntimeException(
                "Seed 파일을 찾을 수 없습니다: {$sqlFile}"
            );
        }

        $sqlContent = file_get_contents($sqlFile);
        if ($sqlContent === false) {
            throw new RuntimeException(
                "Seed 파일을 읽을 수 없습니다: {$sqlFile}"
            );
        }

        // INSERT 문 파싱 및 실행
        $insertStatements = $this->parseInsertStatements($sqlContent);

        foreach ($insertStatements as $insertSql) {
            $connection->exec($insertSql);
        }
    }

    /**
     * 모든 seed 파일을 로드한다.
     *
     * 파일 순서대로 로드되므로, FK 종속성을 고려하여 파일명을 정렬할 것.
     *
     * @param PDO $connection PDO 연결
     */
    public function loadAllSeeds(PDO $connection): void
    {
        $files = scandir($this->fixturesDir);
        if ($files === false) {
            throw new RuntimeException(
                "Fixtures 디렉토리를 읽을 수 없습니다: {$this->fixturesDir}"
            );
        }

        // 숨김 파일과 디렉토리 제거
        $sqlFiles = array_filter($files, function (string $file): bool {
            return !str_starts_with($file, '.') && str_ends_with($file, '.sql');
        });

        // 정렬
        sort($sqlFiles);

        foreach ($sqlFiles as $sqlFile) {
            $tableName = pathinfo($sqlFile, PATHINFO_FILENAME);
            $this->loadSeed($connection, $tableName);
        }
    }

    /**
     * SQL 콘텐츠에서 INSERT 문들을 추출한다.
     *
     * @param string $sqlContent SQL 파일의 전체 내용
     * @return string[] INSERT 문 리스트 (주석 제거, 완전한 문)
     */
    private function parseInsertStatements(string $sqlContent): array
    {
        // 라인 주석 제거 (-- 이후의 모든 내용)
        $lines = explode("\n", $sqlContent);
        $cleanedLines = [];

        foreach ($lines as $line) {
            $line = preg_replace('/--.*$/', '', $line);
            $cleanedLines[] = $line;
        }

        $sqlContent = implode("\n", $cleanedLines);

        // 블록 주석 제거 (/* ... */)
        $sqlContent = preg_replace('/\/\*.*?\*\//s', '', $sqlContent);

        // INSERT 문 추출
        // INSERT INTO ... VALUES (...); 패턴
        $pattern = '/INSERT\s+INTO\s+\w+\s*\([^)]*\)\s*VALUES\s*\([^)]*\)\s*;/i';
        $matches = [];
        preg_match_all($pattern, $sqlContent, $matches);

        return $matches[0];
    }

    /**
     * Fixtures 디렉토리를 반환한다.
     *
     * @return string
     */
    public function getFixturesDir(): string
    {
        return $this->fixturesDir;
    }
}

File: src/persistence/seed_loader.py
"""
Portable seed fixture 로더.

ANSI SQL 기반의 INSERT 문을 파싱하여 데이터베이스에 로드한다.
PostgreSQL과 MariaDB 양쪽에서 동일하게 동작한다.
"""

import re
from pathlib import Path
from typing import Any, Dict, List, Optional

from sqlalchemy.ext.asyncio import AsyncSession
from sqlalchemy.sql import text


class SeedLoader:
    """Portable seed fixture를 로드하는 클래스."""

    def __init__(self, fixtures_dir: Optional[Path] = None):
        """
        초기화.

        Args:
            fixtures_dir: seed 파일들이 있는 디렉토리. 지정하지 않으면 tests/fixtures/seed를 사용.
        """
        if fixtures_dir is None:
            # 기본값: tests/fixtures/seed
            # 이 모듈이 src/persistence에 있으므로, 상대 경로로 접근
            project_root = Path(__file__).parent.parent.parent
            fixtures_dir = project_root / "tests" / "fixtures" / "seed"

        self.fixtures_dir = fixtures_dir

    async def load_seed(self, session: AsyncSession, table_name: str) -> None:
        """
        특정 테이블의 seed 데이터를 로드한다.

        Args:
            session: SQLAlchemy AsyncSession
            table_name: 로드할 테이블명 (파일명으로도 사용: {table_name}.sql)
        """
        sql_file = self.fixtures_dir / f"{table_name}.sql"

        if not sql_file.exists():
            raise FileNotFoundError(f"Seed file not found: {sql_file}")

        # SQL 파일 읽기
        sql_content = sql_file.read_text(encoding="utf-8")

        # INSERT 문 파싱 및 실행
        insert_statements = self._parse_insert_statements(sql_content)

        for insert_sql in insert_statements:
            await session.execute(text(insert_sql))

        await session.commit()

    async def load_all_seeds(self, session: AsyncSession) -> None:
        """
        모든 seed 파일을 로드한다.

        파일 순서대로 로드되므로, FK 종속성을 고려하여 파일명을 정렬할 것.
        """
        seed_files = sorted(self.fixtures_dir.glob("*.sql"))

        for seed_file in seed_files:
            if seed_file.name.startswith("."):
                continue  # 숨김 파일 무시

            table_name = seed_file.stem
            await self.load_seed(session, table_name)

    def _parse_insert_statements(self, sql_content: str) -> List[str]:
        """
        SQL 콘텐츠에서 INSERT 문들을 추출한다.

        Args:
            sql_content: SQL 파일의 전체 내용

        Returns:
            INSERT 문 리스트 (주석 제거, 완전한 문)
        """
        # 주석 제거
        lines = sql_content.split("\n")
        cleaned_lines = []

        for line in lines:
            # SQL 라인 주석 제거 (-- 이후의 모든 내용)
            line = re.sub(r"--.*$", "", line)
            cleaned_lines.append(line)

        sql_content = "\n".join(cleaned_lines)

        # 블록 주석 제거 (/* ... */)
        sql_content = re.sub(r"/\*.*?\*/", "", sql_content, flags=re.DOTALL)

        # INSERT 문 추출
        # INSERT INTO ... VALUES (...); 패턴
        pattern = r"INSERT\s+INTO\s+\w+\s*\([^)]*\)\s*VALUES\s*\([^)]*\)\s*;"

        insert_statements = re.findall(pattern, sql_content, re.IGNORECASE | re.DOTALL)

        return insert_statements
